proc_macros/src/lib.rs

```diff
@@ -165,3 +203,599 @@ pub fn rule_tests(input: TokenStream) -> TokenStream {
+    let rule_instance_per_file_state_field_names = rule_instance_per_file_state_fields
+        .iter()
+        .map(|field| &field.name);
+    let rule_instance_per_file_state_field_initializers = rule_instance_per_file_state_fields
+        .iter()
+        .map(|field| match field.initializer.as_ref() {
+            Some(initializer) => quote!(#initializer),
+            None => quote!(Default::default()),
+        });
+    quote! {
+        impl crate::rule::RuleInstance for #rule_instance_struct_name {
+            fn instantiate_per_file(
+                self: std::sync::Arc<Self>,
+                _file_run_info: &crate::rule::FileRunInfo,
+            ) -> Arc<dyn crate::rule::RuleInstancePerFile> {
+                std::sync::Arc::new(#rule_instance_per_file_struct_name {
+                    rule_instance: self,
+                    #(#rule_instance_per_file_state_field_names: #rule_instance_per_file_state_field_initializers),*
+                })
+            }
+
+            fn rule(&self) -> Arc<dyn crate::rule::Rule> {
+                self.rule.clone()
+            }
+
+            fn listener_queries(&self) -> &[crate::rule::RuleListenerQuery] {
+                &self.listener_queries
+            }
+        }
+    }
+}
+
+fn get_rule_instance_per_file_struct_definition(
+    rule_instance_struct_name: &Ident,
+    rule_instance_per_file_struct_name: &Ident,
+    rule_instance_per_file_state_fields: &[&RuleStateFieldSpec],
+) -> proc_macro2::TokenStream {
+    let state_field_names = rule_instance_per_file_state_fields
+        .iter()
+        .map(|field| &field.name);
+    let state_field_types = rule_instance_per_file_state_fields
+        .iter()
+        .map(|field| &field.type_);
+    quote! {
+        struct #rule_instance_per_file_struct_name {
+            rule_instance: std::sync::Arc<#rule_instance_struct_name>,
+            #(#state_field_names: #state_field_types),*
+        }
+    }
+}
+
+struct SelfAccessRewriter<'a> {
+    rule: &'a Rule,
+}
+
+impl<'a> visit_mut::VisitMut for SelfAccessRewriter<'a> {
+    fn visit_expr_field_mut(&mut self, node: &mut ExprField) {
+        if let Some(self_field_name) = get_self_field_access_name(node) {
+            match self.rule.get_rule_state_scope_for_field(&self_field_name) {
+                Some(RuleStateScope::RuleStatic) => {
+                    let self_field_name = format_ident!("{}", self_field_name);
+                    *node = parse_quote!(self.rule_instance.rule.#self_field_name);
+                    return;
+                }
+                Some(RuleStateScope::PerRun) => {
+                    let self_field_name = format_ident!("{}", self_field_name);
+                    *node = parse_quote!(self.rule.#self_field_name);
+                    return;
+                }
+                _ => (),
+            }
+        }
+        visit_mut::visit_expr_field_mut(self, node);
+    }
+
+    fn visit_expr_macro_mut(&mut self, node: &mut ExprMacro) {
```

Comment:
As if re-writing `self.whatever` expressions wasn't enough, this enables re-writing them when they occur inside "macro arguments" (eg `format!(...)`) - `syn` parses the macro invocation contents as a `TokenStream`, so we try "re-parsing" that as a comma-separated list of `Expr`'s